[MonoAPI] Split type and function headers, add MONO_API_FUNCTION macro

lambdageek opened the pull request:
The idea is that the function header can be included multiple times with
different definitions of `MONO_API_FUNCTION` in order to make it easier to
re-used the definitions for embedding the runtime in late-binding scenarios

Contributes to #64456

lambdageek committed:
[monoapi] Split type and function headers, add MONO_API_FUNCTION macro

The idea is that the function header can be included multiple times with
different definitions of MONO_API_FUNCTION in order to make it easier to
re-used the definitions for embedding the runtime in late-binding scenarios

lambdageek commented:
@joncham @vargaz Want to get your feedback before I do `metadata/` and `mini/`.  

@joncham I think this will work for Unity-Technologies/runtime#4

lambdageek committed:
Split metadata headers too

lambdageek committed:
[monoapi] Use cmake install(DIRECTORY) instead of install(TARGETS)

install(TARGETS) flattens subdirectories  (so all the details/ headers ended up
in the respective parent directory)

am11 commented on src/native/public/mono/utils/details/mono-counters-functions.h:
```diff
@@ -0,0 +1,45 @@
+MONO_API_FUNCTION(MONO_API void, mono_runtime_resource_check_limit,  (int resource_type, uintptr_t value))
+
+
+
```
nit: extra blank lines in couple of new headers can be cleaned up.

lambdageek committed:
Split mini jit headers

One complication here is that the directory is called "jit" outside the
runtime, but "mini", inside.

lambdageek committed:
rename public mono/mini subdirectory to mono/jit

To match how embedders see the tree.

Update the runtime to include <mono/jit/jit.h> instead of <mono/mini/jit.h>.

No change in public API

lambdageek committed:
Split remaining jit header

lambdageek committed:
fix whitespace

lambdageek committed:
Add needed includes

joncham commented:
@lambdageek one thing I just hit in my integration is calling convention. On 32-bit windows CoreCLR is built specifying STDCALL as the calling convention. This required me to decorate all my exported Mono functions with __cdecl. I think the current macro approach will let me do same, just FYI.

lambdageek commented:
> @lambdageek one thing I just hit in my integration is calling convention. On 32-bit windows CoreCLR is built specifying STDCALL as the calling convention. This required me to decorate all my exported Mono functions with __cdecl. I think the current macro approach will let me do same, just FYI.

@joncham, just so I'm clear, you'd do something like this, right?

```cpp
extern "C" {
#define MONO_API_FUNCTION(ret,name,args) typedef ret (__cdecl * name##_type) args;
#include <mono/metadata/details/object-functions.h>
#undef MONO_API_FUNCTION

#define MONO_API_FUNCTION(ret,name,args) static name##_type name;
#include <mono/metadata/details/object-functions.h>
#undef MONO_API_FUNCTION

static void
init_functions()
{
  HMODULE hModule = ...
  #define MONO_API_FUNCTION(ret,name,args) name = (name ##_type)GetProcAddress(hModule, #name );
  #include <mono/metadata/details/object-functions.h>
  #undef MONO_API_FUNCTION
}
```

joncham commented:
> @joncham, just so I'm clear, you'd do something like this, right?

Correct.

akoeplinger commented on src/native/public/mono/jit/mono-private-unstable.h:
```diff
@@ -0,0 +1,25 @@
+ *
+ * Private unstable APIs.
+ *
+ * WARNING: The declarations and behavior of functions in this header are NOT STABLE and can be modified or removed at
```
nit: it would be nice to retain this warning in mono-private-unstable-types.h and mono-private-unstable-functions.h too.

akoeplinger commented on src/native/public/monoapi.cmake:
```diff
@@ -47,25 +58,68 @@ set(metadata_public_headers_base
-  install(TARGETS monoapi_mini PUBLIC_HEADER DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/mono-2.0/mono/jit)
+  install(DIRECTORY mono/utils/ DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/mono-2.0/mono/utils)
+  install(DIRECTORY mono/metadata/ DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/mono-2.0/mono/metadata)
+  install(DIRECTORY mono/jit/ DESTINATION ${CMAKE_INSTALL_INCLUDEDIR}/mono-2.0/mono/jit)
```
this will inadvertently install other files we put into these directories, e.g. README.md, right?
could we instead split `jit_public_headers_base` into the normal and `details/` .h files and `install` them separately here?

lambdageek replied:
> this will inadvertently install other files we put into these directories, e.g. README.md, right? could we instead split `jit_public_headers_base` etc. into the normal and `details/` .h files and `install` them separately here?

Yea I think so. I would just use `install(FILES)` then, I think, right?  I don't really know what i'm doing with cmake install rules.

akoeplinger replied:
Yes that sounds good.

am11 replied:
We normally don't use `install(DIRECTORY`, @lambdageek. [Elsewhere](https://grep.app/search?q=install%28FILES&filter[repo.pattern][0]=dotnet/runtime&filter[lang][0]=CMake), we explicitly capture all the needed files in a variable, then use: `install(FILES ${that_var} DESTINATION path/to/some/dir)`.

lambdageek committed:
add unstable API comment to the details headers, too

lambdageek committed:
use install(FILES) to copy only the headers

not other stray files to the include dir

akoeplinger commented:
We should also update the CODEOWNERS file.

lambdageek committed:
update CODEOWNERS

lambdageek committed:
Remove duplicate definitions of mono_event_get_remove_method

This was already duplicated before the header reorganization,
e.g. here https://github.com/dotnet/runtime/blob/b9a55b4f52243325359ced26e3d4b31ccacdc381/src/native/public/mono/metadata/class.h#L279-L282

lambdageek commented:
I'm going to do a follow-up PR with a slight tweak.

Instead of defining functions like this:
```c
MONO_API_FUNCTION(MONO_API MONO_RT_EXTERNAL_ONLY returnType, mono_function_name, (MonoObject *arg1, MonoObject* arg2))
```

I'm going to pull the `MONO_API` part into the default definition of `MONO_API_FUNCTION`, and probably `MONO_RT_EXTERNAL_ONLY` into a new `modifiers` argument:
```c
MONO_API_FUNCTION(MONO_RT_EXTERNAL_ONLY, returnType, mono_function_name, (MonoObject* arg1, MonoObject *arg2))
```

where we define the macro as:
```c
#define MONO_API_FUNCTION(modifiers, ret, name, args) MONO_API modifiers ret name args;
```


The issue is that `MONO_API` in C++ builds expands to `extern "C"`, which is in an illegal place if you're using a `MONO_API_FUNCTION` to typedef a function pointer type.

I'm less sure about the `modifiers` part - currently `MONO_RT_EXTERNAL_ONLY` is the only one that we have and it expands to the empty string outside the runtime, so maybe it's not worth pulling it out.  But it doesn't really belong with the return type, either.  And there might be some other attributes that we might want in the future (e.g. for deprecation warnings) that different `MONO_API_FUNCTION` definitions might want to strip off.

lambdageek committed:
remove MONO_API modifier from *-functions.h

lambdageek committed:
redefine MONO_API_FUNCTION to include MONO_API

lambdageek commented:
I ended up redefining `MONO_API_FUNCTION` on this PR.  still with 3 arguments.  4 arguments where the 1st argument is mostly empty looked ugly (e.g. `MONO_API_FUNCTION(,MonoObject *, mono_some_function, (MonoObject *arg1))`) and `MONO_RT_EXTERNAL_ONLY` expands to empty outside the runtime build, so it didn't seem worth it just for that.

lambdageek commented:
"CoreCLR Product Build windows arm64 " (checked and release) build failures are https://github.com/dotnet/runtime/pull/65624#issuecomment-1049058533
